### PR TITLE
Docker storage

### PR DESCRIPTION
**Pull request by mvollmer**

### CODE REVIEW

File: pkg/docker/index.html [hunk 1]
```diff
@@ -140,10 +142,10 @@ <h1 translatable="yes">Can't connect to Docker</h1>
...
     <div class="col-md-12 col-lg-4">
       <div class="panel panel-default" id="containers-storage">
-        <div class="storage-label">Storage space</div>
-        <div class="bar-group">
-          <div class="data"></div>
-          <div class="bar"></div>
+        <div class="panel-body">
+          <div id="containers-storage-details">
+          </div>
+          <a translatable="yes" href="#/storage">configure storage...</a>
```
Comment (garrett) [inline pkg/docker/index.html]: I think this should be `translate="yes"` https://www.w3.org/TR/html5/dom.html#the-translate-attribute (same with others below)

File: pkg/docker/index.html [hunk 2]
```diff
@@ -336,6 +338,45 @@ <h1 translatable="yes">Can't connect to Docker</h1>
...
+      <div class="blank-slate-pf">
+        <h1 id="storage-unsupported-message"></h1>
+      </div>
+    </div>
+
+    <div id="storage-details">
+      <div class="col-md-4">
+        <h3 translatable="yes">Overview</h3>
+        <div id="storage-overview"></div>
+        <h3>
```
Comment (garrett) [inline pkg/docker/index.html]: You cannot nest block elements inside an `<h3>` (or any other heading tag).

If you want to fix this easily, just swap the open and close `<div>` for `<span>`. 

[Heading elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements) can only contain [phrasing content](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Phrasing_content) — which, thankfully, includes `<span>`, `<button>`, and others.

Comment (mvollmer) [inline pkg/docker/index.html]: I think the second `<h3>` is bogus here.

File: pkg/docker/index.html (hunk 1 above)
Comment (mvollmer) [inline pkg/docker/index.html]: We use `translatable` because that's what `xgettext` expects for `gettext/glade` files, which we abuse to scan our HTML.  But we should definitely work on moving to `translate`.  Thanks for pointing out that there is a official attribute for this.

File: pkg/docker/overview.js [hunk 3]
```diff
@@ -19,16 +19,18 @@
 
 define([
     "jquery",
+    "react",
```
Comment (dperpeet) [inline pkg/docker/overview.js]: since refactoring, this will need to be `"base1/react",`

File: pkg/docker/storage.jsx [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+define([
+    "jquery",
+    "react",
```
Comment (dperpeet) [inline pkg/docker/storage.jsx]: since refactoring, this will need to be `"base1/react",`

File: pkg/docker/storage.jsx [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+    "base1/cockpit",
+    "storaged-client/client",
+    "storaged-client/utils",
+    "performance/dialog-view"
+], function($, React, cockpit, storage_client, storage_utils, dialog_view) {
+    var _ = cockpit.gettext;
+    var C_ = cockpit.gettext;
+
+    function init_model(callback) {
```
Comment (dperpeet) [inline pkg/docker/storage.jsx]: You don't have to change what is already written, but note that we decided to use camelCase in new js code to match what other projects / React is doing.

File: pkg/docker/storage.jsx [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+                                   checked={self.driveChecked(drive)}/>
+                        </td>
+                        <td><img src="images/drive-harddisk-symbolic.svg"/></td>
+                        <td><span>{drive.name}</span><br/><span>{drive.desc}</span></td>
+                    </tr>);
+            }
+
+            for (d in self.state.checked) {
+                if (!drive_devs[d])
+                    self.state.checked[d] = false;
```
Comment (dperpeet) [inline pkg/docker/storage.jsx]: This is a bit unusual, what's the intent here? self.state *may* be replaced by the React framework, so it's a bit unpredictable, see https://facebook.github.io/react/docs/component-api.html.
If you want to change the state, use `setState`

File: pkg/docker/storage.jsx [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+            }
+
+            function drive_list() {
```
Comment (dperpeet) [inline pkg/docker/storage.jsx]: Why is this a function?
You could just use a variable drive_list and assign the jsx code directly.

File: pkg/docker/storage.jsx [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+                        <td>{storage_utils.fmt_size(drive.Size)}</td>
+                        <td><img src="images/drive-harddisk-symbolic.svg"/></td>
+                        <td>{storage_utils.drive_name(drive)}</td>
+                    </tr>);
+            });
+        }
+
+        dialog_view.show_modal_dialog({ 'title': _("Add Additional Storage"),
+                                        'body':
+                                        <div className="modal-body">
```
Comment (dperpeet) [inline pkg/docker/storage.jsx]: jsx code longer than one line should be wrapped in `( )`

File: pkg/docker/storage.jsx [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+                                                { render_drive_rows() }
+                                            </table>
+                                        </div>,
+                                      },
+                                      { 'actions': [ { 'caption': _("Reformat and add disks"),
+                                                       'clicked': do_it,
+                                                       'style': "danger" } ]
+                                      });
+
+        function do_it() {
```
Comment (dperpeet) [inline pkg/docker/storage.jsx]: I suggest using a more descriptive name here.

File: pkg/docker/storage.jsx [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+                                        'body':
+                                        <div className="modal-body">
+                                            <p>{_("Resetting the storage pool will erase all containers and release disks in the pool.")}</p>
+                                        </div>,
+                                      },
+                                      { 'actions': [ { 'caption': _("Erase containers and reset storage pool"),
+                                                       'clicked': do_it,
+                                                       'style': "danger" } ]
+                                      });
+        function do_it() {
```
Comment (dperpeet) [inline pkg/docker/storage.jsx]: I suggest using a more descriptive name here.

**Comment by dperpeet** [comment]
I added a few comments.

Is there a way to split this pull request somewhat? One way could be to make the storaged changes a separate issue from the docker ones.

File: pkg/docker/index.html (hunk 2 above)
Comment (mvollmer) [inline pkg/docker/index.html]: > I think the second <h3> is bogus here.

No, I was confused.  Fixed by using a `span` instead of the `div`.

@garret could you have a look at the HTML coding in general?  How would you implement your mockup in HTML/CSS?

File: pkg/docker/storage.jsx (hunk 6 above)
Comment (mvollmer) [inline pkg/docker/storage.jsx]: It matters in this scenario:
- check a disk
- hit "add" and confirm the dialog
- the disk disappears from the list and appears in the pool
- hit "reset" and confirm the dialog
- the disk disappears from the pool and appears again in this list

If we don't reset the "checked" state for disks that are not currently in the list, they will be checked when they appear again.

File: pkg/docker/storage.jsx (hunk 7 above)
Comment (mvollmer) [inline pkg/docker/storage.jsx]: Yes, this is a left over from refactoring.  Fixed.

File: pkg/docker/storage.jsx (hunk 9 above)
Comment (mvollmer) [inline pkg/docker/storage.jsx]: Done.

File: pkg/docker/storage.jsx (hunk 10 above)
Comment (mvollmer) [inline pkg/docker/storage.jsx]: Done.

**Comment by mvollmer** [comment]
> Is there a way to split this pull request somewhat? One way could be to make the storaged changes a separate issue from the docker ones.

Sure, but I would like to get a little bit closer to the final version before attempting that.  I might not actually use storaged, for example.

File: pkg/docker/overview.js (hunk 3 above)
Comment (mvollmer) [inline pkg/docker/overview.js]: Rebased and adapted.

File: pkg/docker/storage.jsx [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+                                   checked={self.driveChecked(drive)}/>
+                        </td>
+                        <td><img src="images/drive-harddisk-symbolic.svg"/></td>
+                        <td><span>{drive.name}</span><br/><span>{cockpit.format_bytes(drive.size)} {drive.class}</span></td>
+                    </tr>);
+            }
+
+            for (d in self.state.checked) {
+                if (!drive_paths[d])
+                    self.state.checked[d] = false;
```
Comment (dperpeet) [inline pkg/docker/storage.jsx]: As [commented before](https://github.com/cockpit-project/cockpit/pull/4350#discussion_r62456664), `this.state` shouldn't be modified directly. Change it via `setState` or change the way data flows.

File: pkg/docker/storage.jsx [hunk 12]
```diff
@@ -0,0 +1,440 @@
...
+        },
+
+        driveChecked: function (drive) {
+            return !!this.state.checked[drive.path];
+        },
+        makeDriveToggler: function (drive) {
+            var self = this;
+            return function (event) {
+                self.state.checked[drive.path] = !self.state.checked[drive.path];
+                $(event.currentTarget).find("input").prop("checked", self.state.checked[drive.path]);
```
Comment (dperpeet) [inline pkg/docker/storage.jsx]: This shouldn't be necessary - when the state changes, the component will be updated (re-rendered).

File: pkg/docker/storage.jsx [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+            var self = this;
+            $(this.props.model).on("changed", function () {
+                self.setState({ drives: self.props.model.extra_devices });
+            });
+        },
+
+        driveChecked: function (drive) {
+            return !!this.state.checked[drive.path];
+        },
+        makeDriveToggler: function (drive) {
```
Comment (dperpeet) [inline pkg/docker/storage.jsx]: instead of returning a function, you could take `event` and `drive` as a parameter and bind the function.

Example: https://github.com/cockpit-project/cockpit/blob/master/pkg/base1/cockpit-components-listing.jsx#L162

File: pkg/docker/storage.jsx [hunk 14]
```diff
@@ -0,0 +1,440 @@
...
+    }
+
+    var DriveBox = React.createClass({
+        getInitialState: function () {
+            return {
+                drives: this.props.model.extra_devices,
+                checked: { }
+            };
+        },
+        componentDidMount: function () {
```
Comment (dperpeet) [inline pkg/docker/storage.jsx]: A component will unmount and mount again if it changes. This should at least clean up the `changed` listener on unmount.

How much else from the model is being used? It almost looks like the component could be a lot simpler if instead of the model being passed in, it would just receive a list of devices via props.

File: pkg/docker/storage.jsx [hunk 15]
```diff
@@ -0,0 +1,440 @@
...
+        var dfd = $.Deferred();
+        init_model(function (model) {
+            dfd.resolve(model);
+        });
+
+        model_promise = dfd.promise();
+        return model_promise;
+    }
+
+    var DriveBox = React.createClass({
```
Comment (dperpeet) [inline pkg/docker/storage.jsx]: You should comment on the props going in. Though not supported by react-lite, the propTypes notation can also help: https://github.com/cockpit-project/cockpit/blob/master/pkg/base1/cockpit-components-listing.jsx#L44

File: pkg/docker/storage.jsx (hunk 14 above)
Comment (mvollmer) [inline pkg/docker/storage.jsx]: > How much else from the model is being used?

We need the "change" event.

Comment (mvollmer) [inline pkg/docker/storage.jsx]: > A component will unmount and mount again if it changes. This should at least clean up the changed listener on unmount.

Done.  I wasn't able to get the component to unmount, however...

File: pkg/docker/storage.jsx (hunk 12 above)
Comment (mvollmer) [inline pkg/docker/storage.jsx]: Correct, removed.

File: pkg/docker/storage.jsx (hunk 13 above)
Comment (mvollmer) [inline pkg/docker/storage.jsx]: Good idea, done.

File: pkg/docker/storage.jsx (hunk 11 above)
Comment (mvollmer) [inline pkg/docker/storage.jsx]: I think this isn't actually necessary.  The drives are unticked when the "Add" dialog opens, which should be enough.  If a checked drive disappears and then comes back, we might as well keep it checked.

Comment (mvollmer) [inline pkg/docker/storage.jsx]: So, removed.

**Comment by mvollmer** [comment]
Something is broken about "make distcheck", I guess vpath builds.

**Comment by stefwalter** [comment]
Instead of expecting that cockpit-atomic-storage is on the target system, can we treat it like a script and send it there from the browser? That's analogous to how ansible works, and our current browser driven shell scripts work ... as well as doesn't block running from a container in the future. What do you think?

File: pkg/docker/cockpit-atomic-storage [hunk 16]
```diff
@@ -0,0 +1,482 @@
+#! /usr/bin/python
```
Comment (stefwalter) [inline pkg/docker/cockpit-atomic-storage]: Is this python2 or python3? This is the first part of the Cockpit runtime where we are running python on the target system. We should probably make this script compatible with both, and run it from the browser, rather than install it.

File: pkg/docker/cockpit-atomic-storage [hunk 17]
```diff
@@ -0,0 +1,482 @@
...
+
+def get_devicemapper_thinpool_usage(dev):
+    table_fields = check_output([ "dmsetup", "table", dev ]).split()
+    status_fields = check_output([ "dmsetup", "status", dev ]).split()
+    data_blocksize = int(table_fields[5])
+    data_used = int(status_fields[5].split("/")[0])
+    return data_used*data_blocksize*512
+
+udev_context = pyudev.Context()
+udev_monitor = pyudev.Monitor.from_netlink(udev_context)
```
Comment (stefwalter) [inline pkg/docker/cockpit-atomic-storage]: Can we use ```udevadm monitor``` here instead of pyudev?

File: pkg/docker/cockpit-atomic-storage [hunk 18]
```diff
@@ -0,0 +1,482 @@
...
+    elif d.attributes.asbool("queue/rotational"):
+        return "hdd"
+    else:
+        return "sdd"
+
+can_manage = None
+
+def get_info():
+    drives = { }
+    for d in udev_context.list_devices(subsystem="block"):
```
Comment (stefwalter) [inline pkg/docker/cockpit-atomic-storage]: Can we use ```udevadm info``` here instead of pyudev?

**Comment by stefwalter** [comment]
Added some questions about use of pyudev. The question is whether we can get away with udevadm instead.

File: pkg/docker/storage.jsx [hunk 19]
```diff
@@ -0,0 +1,512 @@
...
+            extra_devices: [ ],
+            total: 0,
+            used: 0
+        };
+
+        function cmp_drive(a, b) {
+            return a.sort_index - b.sort_index;
+        }
+
+        cockpit.spawn([ config.cockpit_atomic_storage, "monitor" ],
```
Comment (stefwalter) [inline pkg/docker/storage.jsx]: When do we kill this process? Should we kill it when the user hides the page? See #4492

**Comment by mvollmer** [comment]
> Is this python2 or python3?

Ah, I see, /usr/bin/python can be either.  I'll test it with both.

**Comment by stefwalter** [comment]
> Ah, I see, /usr/bin/python can be either. I'll test it with both.

Is there a good way to validate this long term? Make sure it doesn't rot? Does testing against various operating systems already result in that?

**Comment by mvollmer** [comment]
> Is there a good way to validate this long term? Make sure it doesn't rot? Does testing against various operating systems already result in that?

I think that's enough.

File: Makefile.am [hunk 20]
```diff
@@ -8,6 +8,7 @@ check_SCRIPTS =
 check_PROGRAMS =
 noinst_SCRIPTS =
 libexec_PROGRAMS =
+libexec_SCRIPTS =
```
Comment (stefwalter) [inline Makefile.am]: I don't think this change is needed any longer.

File: .gitignore [hunk 21]
```diff
@@ -149,3 +149,5 @@ Test*.log
 cockpit-wip.tar*
 Test*.png
 
+/pkg/docker/storage.js
+/pkg/docker/config.json
```
Comment (stefwalter) [inline .gitignore]: I don't think this change is needed any longer.

File: pkg/base1/cockpit.js [hunk 22]
```diff
@@ -3767,6 +3775,13 @@ function basic_scope(cockpit, jquery) {
                 "headers": headers
             });
         };
+
+        self.close = function close(problem) {
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Good idea. But this needs documentation and a unit test.

File: pkg/docker/cockpit-atomic-storage [hunk 23]
```diff
@@ -0,0 +1,501 @@
...
+        with open("/dev/null", "w") as dev_null:
+            can_manage = call([ "atomic", "storage", "--help" ], stdout=dev_null) == 0
+    except:
+        can_manage = False
+
+    old_info = get_info()
+    print_info(old_info)
+
+    mon = subprocess.Popen([ "stdbuf", "-o", "L", "udevadm", "monitor", "-u", "-s", "block"],
+                           bufsize=1, stdout=subprocess.PIPE)
```
Comment (stefwalter) [inline pkg/docker/cockpit-atomic-storage]: Why do we need both bufsize=1 and stdbuf here? Doesn't bufsize=1 turn on line buffering already?

File: pkg/docker/index.html [hunk 24]
```diff
@@ -28,6 +28,8 @@
     <link href="docker.css" type="text/css" rel="stylesheet" id="term-style">
     <script src="../base1/bundle.js"></script>
     <script src="../shell/shell.js"></script>
+    <script src="../storaged-client/bundle.js"></script>
```
Comment (stefwalter) [inline pkg/docker/index.html]: What is the storaged-client here. I don't it defined or refered  to anywhere else.

File: pkg/docker/index.html [hunk 25]
```diff
@@ -336,6 +337,38 @@ <h1 translatable="yes">Can&rsquo;t connect to Docker</h1>
     </div>
   </div>
 
+  <!-- STORAGE -->
+
+  <div id="storage" class="container-fluid c3" hidden>
```
Comment (stefwalter) [inline pkg/docker/index.html]: What is the c3 here?

File: pkg/docker/index.html [hunk 26]
```diff
@@ -28,6 +28,8 @@
     <link href="docker.css" type="text/css" rel="stylesheet" id="term-style">
     <script src="../base1/bundle.js"></script>
     <script src="../shell/shell.js"></script>
+    <script src="../storaged-client/bundle.js"></script>
+    <script src="../performance/bundle.js"></script>
```
Comment (stefwalter) [inline pkg/docker/index.html]: Is this used for something in particular?

File: pkg/docker/images/drive-harddisk-symbolic.svg [hunk 27]
```diff
@@ -0,0 +1,34 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (stefwalter) [inline pkg/docker/images/drive-harddisk-symbolic.svg]: @andreasn What about using http://fontawesome.io/icon/hdd-o/ instead of our own image?

File: pkg/docker/config.json.in [hunk 28]
```diff
@@ -0,0 +1,3 @@
+{
+    "cockpit_atomic_storage": "@libexecdir@/cockpit-atomic-storage"
```
Comment (stefwalter) [inline pkg/docker/config.json.in]: Is this file still needed?

File: pkg/docker/cockpit-atomic-storage (hunk 23 above)
Comment (mvollmer) [inline pkg/docker/cockpit-atomic-storage]: bufsize=1 turns on line buffering in the python process, while stdbuf turns on line buffering in the udevadm process.

Tuning the buffering in the python process is just an optimization. Without it, python does single-byte-reads (`read (fd, buf, 1)`), but with it, it reads bigger chunks.  It doesn't really have to do anything with lines when reading, just with avoiding the single-byte-reads. busize=1 hopefully uses a reasonable default buffer size, and we don't have to choose our own.

Without stdbuf, udevadm itself will buffer its output and not output a event when it happens.  You can see this with
```
strace -e write udevadm monitor | cat
```
vs just
```
strace -e write udevadm monitor
```

I would argue that `udevadm monitor` should always line-buffer its output (or explicitly flush), even when writing to a file.  Anything else is unhelpful.

File: pkg/docker/images/drive-harddisk-symbolic.svg (hunk 27 above)
Comment (andreasn) [inline pkg/docker/images/drive-harddisk-symbolic.svg]: I don't think it looks super-awesome, but could work. I would allow to clean out one more image, so that's a plus for using the font.

Comment (mvollmer) [inline pkg/docker/images/drive-harddisk-symbolic.svg]: If we change this here, we should also change it in `pkg/storaged`.

File: pkg/docker/index.html (hunk 25 above)
Comment (mvollmer) [inline pkg/docker/index.html]: @garrett, do you know?

Comment (mvollmer) [inline pkg/docker/index.html]: No actual CSS rule seems to match it, so I'll remove it.

File: pkg/docker/config.json.in (hunk 28 above)
Comment (mvollmer) [inline pkg/docker/config.json.in]: Removed.

File: pkg/docker/cockpit-atomic-storage (hunk 17 above)
Comment (mvollmer) [inline pkg/docker/cockpit-atomic-storage]: Done.

File: pkg/docker/cockpit-atomic-storage (hunk 18 above)
Comment (mvollmer) [inline pkg/docker/cockpit-atomic-storage]: Done.

File: Makefile.am (hunk 20 above)
Comment (mvollmer) [inline Makefile.am]: Done.

File: .gitignore (hunk 21 above)
Comment (mvollmer) [inline .gitignore]: Removed.

File: pkg/docker/index.html (hunk 24 above)
Comment (mvollmer) [inline pkg/docker/index.html]: Removed.

File: pkg/base1/cockpit.js (hunk 22 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Added.

File: pkg/docker/storage.jsx (hunk 19 above)
Comment (mvollmer) [inline pkg/docker/storage.jsx]: I'll work on this.

**Comment by mvollmer** [comment]
The `cockpit-atomic-storage` process should be stopped when the page is not visible.

**Comment by mvollmer** [comment]
> The cockpit-atomic-storage process should be stopped when the page is not visible.

Done.

**Comment by dperpeet** [comment]
@mvollmer do we want this to go into the next release? if so, this needs a rebase

**Comment by mvollmer** [comment]
> @mvollmer do we want this to go into the next release? if so, this needs a rebase

Done the rebase, and also found a bug: testDevmapperVGroup on fedora-24 should now actually be able to manage the storage, but Cockpit was still refusing to.  Also fixed some unexpected journal message related to restarting docker in the background.

File: pkg/docker/cockpit-atomic-storage (hunk 23 above)
Comment (stefwalter) [inline pkg/docker/cockpit-atomic-storage]: Makes sense. Created a patch upstream too. https://github.com/systemd/systemd/pull/3733
